man/csvstat.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "CSVSTAT" "1" "Aug 16, 2024" "2.2.0" "csvkit"
.SH NAME
csvstat \- csvstat Documentation
.SH DESCRIPTION
.sp
Prints descriptive statistics for all columns in a CSV file. Will intelligently determine the type of each column and then print analysis relevant to that type (ranges for dates, mean and median for integers, etc.):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: csvstat [\-h] [\-d DELIMITER] [\-t] [\-q QUOTECHAR] [\-u {0,1,2,3}] [\-b]
               [\-p ESCAPECHAR] [\-z FIELD_SIZE_LIMIT] [\-e ENCODING] [\-L LOCALE]
               [\-S] [\-\-blanks] [\-\-null\-value NULL_VALUES [NULL_VALUES ...]]
               [\-\-date\-format DATE_FORMAT] [\-\-datetime\-format DATETIME_FORMAT]
               [\-H] [\-K SKIP_LINES] [\-v] [\-l] [\-\-zero] [\-V] [\-\-csv] [\-\-json]
               [\-i INDENT] [\-n] [\-c COLUMNS] [\-\-type] [\-\-nulls] [\-\-non\-nulls]
               [\-\-unique] [\-\-min] [\-\-max] [\-\-sum] [\-\-mean] [\-\-median]
               [\-\-stdev] [\-\-len] [\-\-max\-precision] [\-\-freq]
               [\-\-freq\-count FREQ_COUNT] [\-\-count]
               [\-\-decimal\-format DECIMAL_FORMAT] [\-G] [\-y SNIFF_LIMIT] [\-I]
               [FILE]

Print descriptive statistics for each column in a CSV file.

positional arguments:
  FILE                  The CSV file to operate on. If omitted, will accept
                        input as piped data via STDIN.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-\-csv                 Output results as a CSV table, rather than plain text.
  \-\-json                Output results as JSON text, rather than plain text.
  \-i INDENT, \-\-indent INDENT
                        Indent the output JSON this many spaces. Disabled by
                        default.
  \-n, \-\-names           Display column names and indices from the input CSV
                        and exit.
  \-c COLUMNS, \-\-columns COLUMNS
                        A comma\-separated list of column indices, names or
                        ranges to be examined, e.g. \(dq1,id,3\-5\(dq. Defaults to
                        all columns.
  \-\-type                Only output data type.
  \-\-nulls               Only output whether columns contains nulls.
  \-\-non\-nulls           Only output counts of non\-null values.
  \-\-unique              Only output counts of unique values.
  \-\-min                 Only output smallest values.
  \-\-max                 Only output largest values.
  \-\-sum                 Only output sums.
  \-\-mean                Only output means.
  \-\-median              Only output medians.
  \-\-stdev               Only output standard deviations.
  \-\-len                 Only output the length of the longest values.
  \-\-max\-precision       Only output the most decimal places.
  \-\-freq                Only output lists of frequent values.
  \-\-freq\-count FREQ_COUNT
                        The maximum number of frequent values to display.
  \-\-count               Only output total row count.
  \-\-decimal\-format DECIMAL_FORMAT
                        %\-format specification for printing decimal numbers.
                        Defaults to locale\-specific formatting with \(dq%.3f\(dq.
  \-G, \-\-no\-grouping\-separator
                        Do not use grouping separators in decimal numbers.
  \-y SNIFF_LIMIT, \-\-snifflimit SNIFF_LIMIT
                        Limit CSV dialect sniffing to the specified number of
                        bytes. Specify \(dq0\(dq to disable sniffing entirely, or
                        \(dq\-1\(dq to sniff the entire file.
  \-I, \-\-no\-inference    Disable type inference (and \-\-locale, \-\-date\-format,
                        \-\-datetime\-format, \-\-no\-leading\-zeroes) when parsing
                        the input.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See also: \fI\%Arguments common to all tools\fP\&.
.SH EXAMPLES
.sp
Basic use:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
csvstat examples/realdata/FY09_EDU_Recipients_by_State.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
When an statistic name is passed, only that stat will be printed:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ csvstat \-\-min examples/realdata/FY09_EDU_Recipients_by_State.csv
  1. State Name: None
  2. State Abbreviate: None
  3. Code: 1
  4. Montgomery GI Bill\-Active Duty: 435
  5. Montgomery GI Bill\- Selective Reserve: 48
  6. Dependents\(aq Educational Assistance: 118
  7. Reserve Educational Assistance Program: 60
  8. Post\-Vietnam Era Veteran\(aqs Educational Assistance Program: 1
  9. TOTAL: 768
 10. j: None
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If a single stat \fIand\fP a single column are requested, only a value will be returned:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
$ csvstat \-c 4 \-\-mean examples/realdata/FY09_EDU_Recipients_by_State.csv
6,263.904
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
Christopher Groskopf and contributors
.SH COPYRIGHT
2016, Christopher Groskopf and James McKinney
.\" Generated by docutils manpage writer.
.


CHANGELOG
0.9.2
-----

* Fix version of openpyxl.

0.9.1
-----

* Add Antonio Lima to AUTHORS.
* Add support for ndjson. (#329)
* Add missing docs for csvcut -C. (#227)
* Reorganize docs so TOC works better. (#339)
* Render docs locally with RTD theme.
* Fix header in "tricks" docs.
* Add install instructions to tutorial. (#331)
* Add killer examples to doc index. (#328)
* Reorganize doc index
* Fix broken csvkit module documentation. (#327)
* Fix version of openpyxl to work around encoding issue. (#391, #288)

0.9.0
-----

* Write missing sections of the tutorial. (#32)
* Remove -q arg from sql2csv (conflicts with common flag).
* Fix csvjoin in case where left dataset rows without all columns.
* Rewrote tutorial based on LESO data. (#324)
* Don't error in csvjson if lat/lon columns are null. (#326)
* Maintain field order in output of csvjson.
* Add unit test for json in2csv. (#77)
* Maintain key order when converting JSON into CSV. (#325.)
* Upgrade python-dateutil to version 2.2 (#304)
* Fix sorting of columns with null values. (#302)
* Added release documentation.
* Fill out short rows with null values. (#313)
* Fix unicode output for csvlook and csvstat. (#315)
* Add documentation for --zero. (#323)
* Fix Integrity error when inserting zero rows in database with csvsql. (#299)
* Add Michael Mior to AUTHORS. (#305)
* Add --count option to CSVStat.
* Implement csvformat.
* Fix bug causing CSVKitDictWriter to output 'utf-8' for blank fields.

0.8.0
-----

* Add pnaimoli to AUTHORS.
* Fix column specification in csvstat. (#236)
* Added "Tips and Tricks" documentation. (#297, #298)
* Add Espartaco Palma to AUTHORS.
* Remove unnecessary enumerate calls. (#292)
* Deprecated DBF support for Python 3+.
* Add support for Python 3.3 and 3.4 (#239)

0.7.3
-----

* Fix date handling with openpyxl > 2.0 (#285)
* Add Kristina Durivage to AUTHORS. (#243)
* Added Richard Low to AUTHORS.
* Support SQL queries "directly" on CSV files. (#276)
* Add Tasneem Raja to AUTHORS.
* Fix off-by-one error in open ended column ranges. (#238)
* Add Matt Pettis to AUTHORS.
* Add line numbers flag to csvlook (#244)
* Only install argparse for Python < 2.7. (#224)
* Add Diego Rabatone Oliveira to AUTHORS.
* Add Ryan Murphy to AUTHORS.
* Fix DBF dependency. (#270)

0.7.2
-----

* Fix CHANGELOG for release.

0.7.1
-----

* Fix homepage url in setup.py.

0.7.0
-----

* Fix XLSX datetime normalization bug. (#223)
* Add raistlin7447 to AUTHORS.
* Merged sql2csv utility (#259).
* Add Jeroen Janssens to AUTHORS.
* Validate csvsql DB connections before parsing CSVs. (#257)
* Clarify install process for Ubuntu. (#249)
* Clarify docs for --escapechar. (#242)
* Make ``import csvkit`` API compatible with ``import csv``.
* Update Travis CI link. (#258)
* Add Sébastien Fievet to AUTHORS.
* Use case-sensitive name for SQLAlchemy (#237)
* Add Travis Swicegood to AUTHORS.

0.6.1
-----

* Add Chris Rosenthal to AUTHORS.
* Fix multi-file input to csvsql. (#193)
* Passing --snifflimit=0 to disable dialect sniffing. (#190)
* Add aarcro to the AUTHORS file.
* Improve performance of csvgrep. (#204)
* Add Matt Dudys to AUTHORS.
* Add support for --skipinitialspace. (#201)
* Add Joakim Lundborg to AUTHORS.
* Add --no-inference option to in2csv and csvsql. (#206)
* Add Federico Scrinzi to AUTHORS file.
* Add --no-header-row to all tools. (#189)
* Fix csvstack blowing up on empty files. (#209)
* Add Chris Rosenthal to AUTHORS file.
* Add --db-schema option to csvsql. (#216)
* Add Shane StClair to AUTHORS file.
* Add --no-inference support to csvsort. (#222)

0.5.0
-----

* Implement geojson support in csvjson. (#159)
* Optimize writing of eight bit codecs. (#175)
* Created csvpy. (#44)
* Support --not-columns for excluding columns. (#137)
* Add Jan Schulz to AUTHORS file.
* Add Windows scripts. (#111, #176)
* csvjoin, csvsql and csvstack will no longer hold open all files. (#178)
* Added Noah Hoffman to AUTHORS.
* Make csvlook output compatible with emacs table markup. (#174)

0.4.4
-----

* Add Derek Wilson to AUTHORS.
* Add Kevin Schaul to AUTHORS.
* Add DBF support to in2csv. (#11, #160)
* Support --zero option for zero-based column indexing. (#144)
* Support mixing nulls and blanks in string columns.
* Add --blanks option to csvsql. (#149)
* Add multi-file (glob) support to csvsql. (#146)
* Add Gregory Temchenko to AUTHORS.
* Add --no-create option to csvsql. (#148)
* Add Anton Ian Sipos to AUTHORS.
* Fix broken pipe errors. (#150)

0.4.3
-----

* Begin CHANGELOG (a bit late, I'll admit).


csvkit/utilities/csvstat.py
#!/usr/bin/env python

import json
import locale
import warnings
from collections import Counter, OrderedDict
from decimal import Decimal

import agate

from csvkit.cli import CSVKitUtility, default_float_decimal, parse_column_identifiers

locale.setlocale(locale.LC_ALL, '')
OPERATIONS = OrderedDict([
    ('type', {
        'aggregation': None,
        'label': 'Type of data: ',
    }),
    ('nulls', {
        'aggregation': agate.HasNulls,
        'label': 'Contains null values: ',
    }),
    ('nonnulls', {
        'aggregation': agate.Count,
        'label': 'Non-null values: ',
    }),
    ('unique', {
        'aggregation': None,
        'label': 'Unique values: ',
    }),
    ('min', {
        'aggregation': agate.Min,
        'label': 'Smallest value: ',
    }),
    ('max', {
        'aggregation': agate.Max,
        'label': 'Largest value: ',
    }),
    ('sum', {
        'aggregation': agate.Sum,
        'label': 'Sum: ',
    }),
    ('mean', {
        'aggregation': agate.Mean,
        'label': 'Mean: ',
    }),
    ('median', {
        'aggregation': agate.Median,
        'label': 'Median: ',
    }),
    ('stdev', {
        'aggregation': agate.StDev,
        'label': 'StDev: ',
    }),
    ('len', {
        'aggregation': agate.MaxLength,
        'label': 'Longest value: ',
    }),
    ('maxprecision', {
        'aggregation': agate.MaxPrecision,
        'label': 'Most decimal places: ',
    }),
    ('freq', {
        'aggregation': None,
        'label': 'Most common values: ',
    }),
])


class CSVStat(CSVKitUtility):
    description = 'Print descriptive statistics for each column in a CSV file.'

    def add_arguments(self):
        self.argparser.add_argument(
            '--csv', dest='csv_output', action='store_true',
            help='Output results as a CSV table, rather than plain text.')
        self.argparser.add_argument(
            '--json', dest='json_output', action='store_true',
            help='Output results as JSON text, rather than plain text.')
        self.argparser.add_argument(
            '-i', '--indent', dest='indent', type=int,
            help='Indent the output JSON this many spaces. Disabled by default.')
        self.argparser.add_argument(
            '-n', '--names', dest='names_only', action='store_true',
            help='Display column names and indices from the input CSV and exit.')
        self.argparser.add_argument(
            '-c', '--columns', dest='columns',
            help='A comma-separated list of column indices, names or ranges to be examined, e.g. "1,id,3-5". '
                 'Defaults to all columns.')
        self.argparser.add_argument(
            '--type', dest='type_only', action='store_true',
            help='Only output data type.')
        self.argparser.add_argument(
            '--nulls', dest='nulls_only', action='store_true',
            help='Only output whether columns contains nulls.')
        self.argparser.add_argument(
            '--non-nulls', dest='nonnulls_only', action='store_true',
            help='Only output counts of non-null values.')
        self.argparser.add_argument(
            '--unique', dest='unique_only', action='store_true',
            help='Only output counts of unique values.')
        self.argparser.add_argument(
            '--min', dest='min_only', action='store_true',
            help='Only output smallest values.')
        self.argparser.add_argument(
            '--max', dest='max_only', action='store_true',
            help='Only output largest values.')
        self.argparser.add_argument(
            '--sum', dest='sum_only', action='store_true',
            help='Only output sums.')
        self.argparser.add_argument(
            '--mean', dest='mean_only', action='store_true',
            help='Only output means.')
        self.argparser.add_argument(
            '--median', dest='median_only', action='store_true',
            help='Only output medians.')
        self.argparser.add_argument(
            '--stdev', dest='stdev_only', action='store_true',
            help='Only output standard deviations.')
        self.argparser.add_argument(
            '--len', dest='len_only', action='store_true',
            help='Only output the length of the longest values.')
        self.argparser.add_argument(
            '--max-precision', dest='maxprecision_only', action='store_true',
            help='Only output the most decimal places.')
        self.argparser.add_argument(
            '--freq', dest='freq_only', action='store_true',
            help='Only output lists of frequent values.')
        self.argparser.add_argument(
            '--freq-count', dest='freq_count', type=int,
            help='The maximum number of frequent values to display.')
        self.argparser.add_argument(
            '--count', dest='count_only', action='store_true',
            help='Only output total row count.')
        self.argparser.add_argument(
            '--decimal-format', dest='decimal_format', type=str, default='%.3f',
            help='%%-format specification for printing decimal numbers. '
                 'Defaults to locale-specific formatting with "%%.3f".')
        self.argparser.add_argument(
            '-G', '--no-grouping-separator', dest='no_grouping_separator', action='store_true',
            help='Do not use grouping separators in decimal numbers.')
        self.argparser.add_argument(
            '-y', '--snifflimit', dest='sniff_limit', type=int, default=1024,
            help='Limit CSV dialect sniffing to the specified number of bytes. '
                 'Specify "0" to disable sniffing entirely, or "-1" to sniff the entire file.')
        self.argparser.add_argument(
            '-I', '--no-inference', dest='no_inference', action='store_true',
            help='Disable type inference when parsing the input. Disable reformatting of values.')

    def main(self):
        if self.args.names_only:
            self.print_column_names()
            return

        if self.additional_input_expected():
            self.argparser.error('You must provide an input file or piped data.')

        operations = [op for op in OPERATIONS.keys() if getattr(self.args, op + '_only')]

        if len(operations) > 1:
            self.argparser.error('Only one operation argument may be specified (--mean, --median, etc).')

        if operations and self.args.csv_output:
            self.argparser.error(
                'You may not specify --csv and an operation (--mean, --median, etc) at the same time.')
        if operations and self.args.json_output:
            self.argparser.error(
                'You may not specify --json and an operation (--mean, --median, etc) at the same time.')
        if operations and self.args.count_only:
            self.argparser.error(
                'You may not specify --count and an operation (--mean, --median, etc) at the same time.')

        if self.args.count_only:
            count = len(list(agate.csv.reader(self.skip_lines(), **self.reader_kwargs)))

            if not self.args.no_header_row:
                count -= 1

            self.output_file.write('%i\n' % count)

            return

        sniff_limit = self.args.sniff_limit if self.args.sniff_limit != -1 else None
        table = agate.Table.from_csv(
            self.input_file,
            skip_lines=self.args.skip_lines,
            sniff_limit=sniff_limit,
            column_types=self.get_column_types(),
            **self.reader_kwargs,
        )

        column_ids = parse_column_identifiers(
            self.args.columns,
            table.column_names,
            self.get_column_offset(),
        )

        kwargs = {}

        if self.args.freq_count:
            kwargs['freq_count'] = self.args.freq_count

        # Output a single stat
        if operations:
            if len(column_ids) == 1:
                self.print_one(table, column_ids[0], operations[0], label=False, **kwargs)
            else:
                for column_id in column_ids:
                    self.print_one(table, column_id, operations[0], **kwargs)
        else:
            stats = {}

            for column_id in column_ids:
                stats[column_id] = self.calculate_stats(table, column_id, **kwargs)

            if self.args.csv_output:
                self.print_csv(table, column_ids, stats)
            elif self.args.json_output:
                self.print_json(table, column_ids, stats)
            else:
                self.print_stats(table, column_ids, stats)

    def is_finite_decimal(self, value):
        return isinstance(value, Decimal) and value.is_finite()

    def _calculate_stat(self, table, column_id, op_name, op_data, **kwargs):
        getter = globals().get(f'get_{op_name}')

        with warnings.catch_warnings():
            warnings.simplefilter('ignore', agate.NullCalculationWarning)

            try:
                if getter:
                    return getter(table, column_id, **kwargs)

                op = op_data['aggregation']
                v = table.aggregate(op(column_id))

                if self.is_finite_decimal(v) and not self.args.json_output:
                    return format_decimal(v, self.args.decimal_format, self.args.no_grouping_separator)

                return v
            except Exception:
                pass

    def print_one(self, table, column_id, op_name, label=True, **kwargs):
        """
        Print data for a single statistic.
        """
        column_name = table.column_names[column_id]

        stat = self._calculate_stat(table, column_id, op_name, OPERATIONS[op_name], **kwargs)

        # Formatting
        if op_name == 'freq':
            stat = ', '.join([f"\"{str(row['value'])}\": {row['count']}" for row in stat])
            stat = '{ %s }' % stat

        if label:
            self.output_file.write('%3i. %s: %s\n' % (column_id + 1, column_name, stat))
        else:
            self.output_file.write(f'{stat}\n')

    def calculate_stats(self, table, column_id, **kwargs):
        """
        Calculate stats for all valid operations.
        """
        return {
            op_name: self._calculate_stat(table, column_id, op_name, op_data, **kwargs)
            for op_name, op_data in OPERATIONS.items()
        }

    def print_stats(self, table, column_ids, stats):
        """
        Print data for all statistics.
        """
        label_column_width = max(len(op_data['label']) for op_data in OPERATIONS.values())

        for column_id in column_ids:
            column_name = table.column_names[column_id]
            column = table.columns[column_id]
            column_stats = stats[column_id]

            self.output_file.write('%3i. "%s"\n\n' % (column_id + 1, column_name))

            for op_name, op_data in OPERATIONS.items():
                if column_stats[op_name] is None:
                    continue

                label = '{label:{label_column_width}}'.format(**{
                    'label_column_width': label_column_width,
                    'label': op_data['label'],
                })

                if op_name == 'freq':
                    for i, row in enumerate(column_stats['freq']):
                        if i == 0:
                            self.output_file.write(f'\t{label} ')
                        else:
                            self.output_file.write('\t{label:{label_column_width}} '.format(**{
                                'label_column_width': label_column_width,
                                'label': '',
                            }))

                        if isinstance(column.data_type, agate.Number):
                            v = row['value']

                            if self.is_finite_decimal(v):
                                v = format_decimal(v, self.args.decimal_format, self.args.no_grouping_separator)
                        else:
                            v = str(row['value'])

                        self.output_file.write(f"{v} ({row['count']}x)\n")

                    continue

                v = column_stats[op_name]

                if op_name == 'nulls' and v:
                    v = f'{v} (excluded from calculations)'
                elif op_name == 'len':
                    v = f'{v} characters'

                self.output_file.write(f'\t{label} {v}\n')

            self.output_file.write('\n')

        self.output_file.write(f'Row count: {len(table.rows)}\n')

    def print_csv(self, table, column_ids, stats):
        """
        Print data for all statistics as a CSV table.
        """
        header = ['column_id', 'column_name'] + list(OPERATIONS)

        writer = agate.csv.DictWriter(self.output_file, fieldnames=header)
        writer.writeheader()

        for row in self._rows(table, column_ids, stats):
            if 'freq' in row:
                row['freq'] = ', '.join([str(row['value']) for row in row['freq']])
            writer.writerow(row)

    def print_json(self, table, column_ids, stats):
        """
        Print data for all statistics as a JSON text.
        """
        data = list(self._rows(table, column_ids, stats))

        json.dump(data, self.output_file, default=default_float_decimal, ensure_ascii=False, indent=self.args.indent)

    def _rows(self, table, column_ids, stats):
        for column_id in column_ids:
            column_name = table.column_names[column_id]
            column_stats = stats[column_id]

            output_row = {'column_id': column_id + 1, 'column_name': column_name}
            for op_name, _op_data in OPERATIONS.items():
                if column_stats[op_name] is not None:
                    output_row[op_name] = column_stats[op_name]

            yield output_row


def format_decimal(d, f='%.3f', no_grouping_separator=False):
    return locale.format_string(f, d, grouping=not no_grouping_separator).rstrip('0').rstrip('.')


# These are accessed via: globals().get(f'get_{op_name}')
def get_type(table, column_id, **kwargs):
    return f'{table.columns[column_id].data_type.__class__.__name__}'


def get_unique(table, column_id, **kwargs):
    return len(table.columns[column_id].values_distinct())


def get_freq(table, column_id, freq_count=5, **kwargs):
    values = table.columns[column_id].values()
    return [
        {'value': r[0], 'count': r[1]}
        for r in Counter(values).most_common(freq_count)
    ]


def launch_new_instance():
    utility = CSVStat()
    utility.run()


if __name__ == '__main__':
    launch_new_instance()
